### ISSUE TITLE
Restore and forget snapshot stored on Artifactory via rclone WebDAV backend does not work

### ISSUE THREAD

**Issue by kohms** [restic/restic#3936]
Hi,
I am not really sure, if my problem is a limitation of Artifactory's WebDAV implementation, a bug in rclone or in restic. If this is not the right location, I am sorry, but I wanted to report the problem as I would search for issues on this repo, if I am looking for a known issue.

Output of `restic version`
--------------------------
```bash
$ restic version
restic 0.14.0 compiled with go1.19 on linux/amd64

$ rclone self-update
2022/09/19 08:20:52 NOTICE: rclone is up to date
$ rclone version
rclone v1.59.2
- os/version: ubuntu 22.04 (64 bit)
- os/kernel: 5.15.0-47-generic (x86_64)
- os/type: linux
- os/arch: amd64
- go/version: go1.18.6
- go/linking: static
- go/tags: none
```

The Artifactory version reported from https://xxx.jfrog.io/artifactory/api/system/version is:
```
{
  "version" : "7.42.4",
  "revision" : "74204900",
  "addons" : [ "aol", "ha", "build", "docker", "vagrant", "replication", "filestore", "plugins", "gems", "composer", "npm", "bower", "git-lfs", "nuget", "debian", "opkg", "rpm", "cocoapods", "conan", "vcs", "pypi", "release-bundle", "jf-connect", "jf-event", "replicator", "keys", "alpine", "cargo", "chef", "federated", "git", "observability", "pub", "rest", "swift", "conda", "terraform", "tracker", "license", "puppet", "ldap", "sso", "layouts", "properties", "search", "securityresourceaddon", "filtered-resources", "p2", "watch", "webstart", "support", "xray", "retention" ],
  "license" : "xxx",
  "entitlements" : {
    "EVENT_BASED_PULL_REPLICATION" : true,
    "SMART_REMOTE_TARGET_FOR_EDGE" : false,
    "REPO_REPLICATION" : true,
    "MULTIPUSH_REPLICATION" : true
  }
}
```

How did you run restic exactly?
-------------------------------
I tried to store my backups on Artifactory via its WebDAV integration using rclone: <https://www.jfrog.com/confluence/display/JFROG/Using+WebDAV>

I initialized the repo, backed up data and tried to restore a given snapshot id.
Restoring or forgetting a particular snapshot id did not work, restoring latest or using the fuse mount worked as expected.

What backend/server/service did you use to store the repository?
----------------------------------------------------------------
I used a free Artifactory Cloud (Trail) via its WebDAV integration using rclone.

Expected behavior
-----------------
I expected to be able to restore and forget a given snapshot id like for the rest server or onedrive or sftp backend.

Actual behavior
---------------
I get this error message:

```bash
$ DEBUG_LOG=restic-restore-debug.log restic restore 18977eee -r rclone:artifactory:/restic --target /tmp/restore/
debug log file restic-restore-debug.log
debug enabled
enter password for repository: 
repository 23ae2878 opened (repository version 2) successfully, password is correct
invalid id "18977eee": multiple IDs with prefix "18977eee" found
```

Steps to reproduce the behavior
-------------------------------

I created a cloud trial at <https://jfrog.com/start-free/#saas>, created a new Identity Token (Profile > Generate an Identity Token) and configured rclone:

```bash
$ rclone config
Current remotes:

Name                 Type
====                 ====
onedrive             onedrive

e) Edit existing remote
n) New remote
d) Delete remote
r) Rename remote
c) Copy remote
s) Set configuration password
q) Quit config
e/n/d/r/c/s/q> n
name> artifactory
Type of storage to configure.
Enter a string value. Press Enter for the default ("").
Choose a number from below, or type in your own value
 1 / 1Fichier
   \ "fichier"
 2 / Alias for an existing remote
   \ "alias"
 3 / Amazon Drive
   \ "amazon cloud drive"
 4 / Amazon S3 Compliant Storage Provider (AWS, Alibaba, Ceph, Digital Ocean, Dreamhost, IBM COS, Minio, Tencent COS, etc)
   \ "s3"
 5 / Backblaze B2
   \ "b2"
 6 / Box
   \ "box"
 7 / Cache a remote
   \ "cache"
 8 / Citrix Sharefile
   \ "sharefile"
 9 / Dropbox
   \ "dropbox"
10 / Encrypt/Decrypt a remote
   \ "crypt"
11 / FTP Connection
   \ "ftp"
12 / Google Cloud Storage (this is not Google Drive)
   \ "google cloud storage"
13 / Google Drive
   \ "drive"
14 / Google Photos
   \ "google photos"
15 / Hubic
   \ "hubic"
16 / In memory object storage system.
   \ "memory"
17 / Jottacloud
   \ "jottacloud"
18 / Koofr
   \ "koofr"
19 / Local Disk
   \ "local"
20 / Mail.ru Cloud
   \ "mailru"
21 / Microsoft Azure Blob Storage
   \ "azureblob"
22 / Microsoft OneDrive
   \ "onedrive"
23 / OpenDrive
   \ "opendrive"
24 / OpenStack Swift (Rackspace Cloud Files, Memset Memstore, OVH)
   \ "swift"
25 / Pcloud
   \ "pcloud"
26 / Put.io
   \ "putio"
27 / SSH/SFTP Connection
   \ "sftp"
28 / Sugarsync
   \ "sugarsync"
29 / Transparently chunk/split large files
   \ "chunker"
30 / Union merges the contents of several upstream fs
   \ "union"
31 / Webdav
   \ "webdav"
32 / Yandex Disk
   \ "yandex"
33 / http Connection
   \ "http"
34 / premiumize.me
   \ "premiumizeme"
35 / seafile
   \ "seafile"
Storage> webdav
** See help for webdav backend at: https://rclone.org/webdav/ **

URL of http host to connect to
Enter a string value. Press Enter for the default ("").
Choose a number from below, or type in your own value
 1 / Connect to example.com
   \ "https://example.com"
url> https://xxx.jfrog.io/artifactory/example-repo-local
Name of the Webdav site/service/software you are using
Enter a string value. Press Enter for the default ("").
Choose a number from below, or type in your own value
 1 / Nextcloud
   \ "nextcloud"
 2 / Owncloud
   \ "owncloud"
 3 / Sharepoint
   \ "sharepoint"
 4 / Other site/service or software
   \ "other"
vendor> other
User name
Enter a string value. Press Enter for the default ("").
user> xxx@xxx.xx
Password.
y) Yes type in my own password
g) Generate random password
n) No leave this optional password blank (default)
y/g/n> y
Enter the password:
password:
Confirm the password:
password:
Bearer token instead of user/pass (eg a Macaroon)
Enter a string value. Press Enter for the default ("").
bearer_token> 
Edit advanced config? (y/n)
y) Yes
n) No (default)
y/n> 
Remote config
--------------------
[artifactory]
url = https://xxx.jfrog.io/artifactory/example-repo-local
vendor = other
user = xxx@xxx.xx
pass = *** ENCRYPTED ***
--------------------
y) Yes this is OK (default)
e) Edit this remote
d) Delete this remote
y/e/d> 
Current remotes:

Name                 Type
====                 ====
artifactory          webdav
onedrive             onedrive

e) Edit existing remote
n) New remote
d) Delete remote
r) Rename remote
c) Copy remote
s) Set configuration password
q) Quit config
e/n/d/r/c/s/q> q
```

Afterwards, I initialized a new repo and checked via rclone, if directories were created which looks okay.
Please find all the debug logs in [restic-debug-logs.zip](https://github.com/restic/restic/files/9596714/restic-debug-logs.zip).

```bash
$ DEBUG_LOG=restic-init-debug.log restic init -r rclone:artifactory:/restic
debug log file restic-init-debug.log
debug enabled
enter password for new repository: 
enter password again: 
created restic repository 23ae2878b3 at rclone:artifactory:/restic

Please note that knowledge of your password is required to access
the repository. Losing your password means that your data is
irrecoverably lost.
$ rclone lsd artifactory:/
          -1 2022-09-19 08:24:42        -1 restic
$ rclone lsd artifactory:/restic
          -1 2022-09-19 08:24:43        -1 data
          -1 2022-09-19 08:24:43        -1 index
          -1 2022-09-19 08:24:43        -1 keys
          -1 2022-09-19 08:24:43        -1 locks
          -1 2022-09-19 08:24:43        -1 snapshots
```

Running a backup seem to work okay as well:

```bash
$ DEBUG_LOG=restic-backup-debug.log restic backup -r rclone:artifactory:/restic /tmp/restic-init-debug.log 
debug log file restic-backup-debug.log
debug enabled
enter password for repository: 
repository 23ae2878 opened (repository version 2) successfully, password is correct
created new cache in /home/xxx/.cache/restic
no parent snapshot found, will read all files

Files:           1 new,     0 changed,     0 unmodified
Dirs:            1 new,     0 changed,     0 unmodified
Added to the repository: 6.039 KiB (2.034 KiB stored)

processed 1 files, 5.298 KiB in 0:02
snapshot 18977eee saved

$ restic snapshots -r rclone:artifactory:/restic
enter password for repository: 
repository 23ae2878 opened (repository version 2) successfully, password is correct
ID        Time                 Host        Tags        Paths
---------------------------------------------------------------------------------
18977eee  2022-09-19 08:26:35  xxx                    /tmp/restic-init-debug.log
---------------------------------------------------------------------------------
1 snapshots


$ rclone lsd artifactory:/restic
          -1 2022-09-19 08:24:43        -1 data
          -1 2022-09-19 08:24:43        -1 index
          -1 2022-09-19 08:24:43        -1 keys
          -1 2022-09-19 08:24:43        -1 locks
          -1 2022-09-19 08:24:43        -1 snapshots

$ rclone lsd artifactory:/restic/data
          -1 2022-09-19 08:26:42        -1 2d
          -1 2022-09-19 08:26:42        -1 8c

```

But, if I try to forget or restore a particular snapshot by ID, I get those error messages:

```bash
$ DEBUG_LOG=restic-prune-debug.log restic forget --prune 18977eee -r rclone:artifactory:/restic
debug log file restic-prune-debug.log
debug enabled
enter password for repository: 
repository 23ae2878 opened (repository version 2) successfully, password is correct
Ignoring "18977eee": multiple IDs with prefix "18977eee" found

$ DEBUG_LOG=restic-restore-debug.log restic restore 18977eee -r rclone:artifactory:/restic --target /tmp/restore/
debug log file restic-restore-debug.log
debug enabled
enter password for repository: 
repository 23ae2878 opened (repository version 2) successfully, password is correct
invalid id "18977eee": multiple IDs with prefix "18977eee" found
```

Interesting is, that I can restore, if I use "latest" instead of the snapshot id:

```bash
$ DEBUG_LOG=restic-restore-debug.log restic restore latest -r rclone:artifactory:/restic --target /tmp/restore/
debug log file restic-restore-debug.log
debug enabled
enter password for repository: 
repository 23ae2878 opened (repository version 2) successfully, password is correct
restoring <Snapshot 18977eee of [/tmp/restic-init-debug.log] at 2022-09-19 08:26:35.880118336 +0200 CEST by xxx@xxx> to /tmp/restore/
```

I can also use the fuse mount and can even go via the snapshots folder, but that will print a timestamp and not the snapshot id, maybe that is implemented differently internally in restic.
```
$ restic mount /tmp/mount -r rclone:artifactory:/restic
enter password for repository: 
repository 23ae2878 opened (repository version 2) successfully, password is correct
Now serving the repository at /tmp/mount
Use another terminal or tool to browse the contents of this folder.
When finished, quit with Ctrl-c here or umount the mountpoint.
  signal interrupt received, cleaning up
```

Do you have any idea what may have caused this?
-----------------------------------------------
Not sure, but <https://www.jfrog.com/confluence/display/JFROG/Using+WebDAV> mentiones some problems with locking, maybe this is used internally in restic?

Do you have an idea how to solve the issue?
-------------------------------------------
Unfortunately, not.

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Yes, I used restic to backup my photo collection to onedrive from my NAS which was quite cool as restic + rclone worked natively without having to run it via docker.

**Comment by kohms** [comment]
Please find the related discussion also in https://forum.restic.net/t/which-protocol-should-i-choose-for-remote-linux-backups/5446

**Comment by kohms** [comment]
Updates from the discussion:
Artifactory seem to calculate a sha256 checksum for each uploaded file. Those checksums are not visible by default, if accessed via WebUI, but if WebDAV is used, those checksums are displayed as additional files.
I think that is a bug in Artifactory and not in restic, therefore I opened https://www.jfrog.com/jira/browse/RTFACT-27453.

Restic could workaround the mentioned problem by not including any filenames with endings in the snapshot directory, but I want to check first, if JFrog accepts it as a bug and will stop injecting new files into folders in the WebDAV view.

**Comment by kohms** [comment]
See https://github.com/restic/restic/pull/3943

Thanks a lot 🙂

**Comment by MichaelEischer** [comment]
Fixed by #3943.

**Comment by kohms** [comment]
Just tried to compile the fix manually and it prefectly resolves the issue, thanks 🙂

For reference, I posted the output in the forum https://forum.restic.net/t/which-protocol-should-i-choose-for-remote-linux-backups/5446/38?u=konrad
